docs(docs-infra): traducción del archivo router-tutorial.md (#385)
angular-hispano/angular: aio/content/guide/router-tutorial.md

# Usando rutas de Angular en una aplicación de una sola página

Este tutorial explica cómo puedes construir una SPA que utilice multiples rutas en Angular.

En una SPA, todas las funcionalidades de la aplicación existen en una sola página de HTML.
A medida que los usuarios acceden a las funcionalidades de tu aplicación, el navegador necesita representar solo las partes que son importantes para el usuario, en lugar de cargar una nueva página por cada acción. Este patrón puede mejorar significativamente la experiencia de usuario de tu aplicación.

Utiliza las rutas para definir la manera en la que los usuarios navegan a través de tu aplicación.
Puedes agregar rutas para definir cómo los usuarios navegan en tu aplicación de una parte a otra. También puedes configurar rutas para protegerte contra comportamientos inesperados o no autorizados.

> Para ver en funcionamiento una aplicación de ejemplo basado en este tutorial puede consultar: <live-example></live-example>.

Reviewer: ```suggestion
Para ver en funcionamiento una aplicación de ejemplo basado en este tutorial puedes consultar: <live-example></live-example>.
```

## Objetivos

- Organizar las funcionalidades de una aplicación ejemplo en módulos.
- Definir como se navega hacia un componente.
- Pasar información hacia un componente a través de parámetros.
- Estructurar rutas mediante el anidamiento de varias rutas
- Comprobar si los usuarios pueden acceder a cierta ruta en particular.
- Controlar si la aplicación puede descartar los cambios no guardados.
- Mejorar el desempeño por medio de la obtención previa de datos de ruta y módulos con funcionalidades de carga diferida.
- Requerir criterios específicos para cargar componentes.

## Pre-requisitos

- HTML
- CSS
- JavaScript
- [Angular CLI](/cli)

> Puede que te sea util el [tutorial de Tour of Heroes](/tutorial), pero es opcional.

Reviewer: ```suggestion
Puede que te sea útil el [tutorial de Tour de Héroes](/tutorial), pero es opcional.
```

1. Crea el segundo componente _heroes-list_ usando el siguiente comando:

 <code-example language="sh">
 ng generate component heroes-list
</code-example>

1. En tu editor de código localiza el archivo llamado `heroes-list.component.html` y remplaza el contenido de _placeholder_ con la siguiente linea de HTML:

   <code-example header="src/app/heroes-list/heroes-list.component.html" path="router-tutorial/src/app/heroes-list/heroes-list.component.html"></code-example>

1. En tu editor de código abre el archivo llamado `app.component.html` y remplaza su contenido con la siguiente linea de HTML:

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="setup"></code-example>

1. Verifica que al ejecutar el siguiente comando `ng serve` en la terminal, la aplicación se ejecute adecuadamente.

<code-example language="sh">
   ng serve
  </code-example>

1. Ahora abre el navegador en `http://localhost:4200`.
   Deberías ver una sola página web, que consta de un título y el HTML de sus dos componentes.

## Pasos para importar `RouterModule` from `@angular/router`

> El enrutamiento le permite mostrar vistas específicas de su aplicación según la ruta de la URL.

Reviewer: ```suggestion
El enrutamiento te permite mostrar vistas específicas de su aplicación según la ruta de la URL.
```

> Para agregar esta funcionalidad a tu aplicación de ejemplo, necesitas actualizar el archivo `app.module.ts` para usar el módulo,` RouterModule`.

Reviewer: ```suggestion
Para agregar esta funcionalidad a tu aplicación de ejemplo, necesitas actualizar el archivo `app.module.ts` para usar el módulo ` RouterModule`.
```

Importa este módulo desde `@angular/router`.

1. Desde tu editor de código, abre el archivo `app.module.ts`.
1. Agrega la siguiente declaración de `import`
   <code-example header="src/app/app.module.ts" path="router-tutorial/src/app/app.module.ts" region="router-import"></code-example>

## Define tus rutas

En esta sección definirás dos rutas:

- La ruta `/crisis-center` que direcionará hacia el componente `crisis-center`.
- La ruta `/heroes-list` que direcionará hacia el componente `heroes-list`.

> Una definición de ruta es un objeto JavaScript. Cada ruta tiene normalmente dos propiedades. La primera propiedad es `path` qué es una cadena de caracteres que especifica la ruta URL de la ruta. La segunda propiedad es `component` qué es una cadena de caracteres que especifica el componente que debe mostrar su aplicación para esa ruta.

Reviewer: ```suggestion
Una definición de ruta es un objeto JavaScript. Cada ruta tiene normalmente dos propiedades. La primera propiedad es `path` que es una cadena de caracteres que especifica la ruta URL de la ruta. La segunda propiedad es `component` que es una cadena de caracteres que especifica el componente que debe mostrar su aplicación para esa ruta.
```

1. Desde tu editor de código abre el archivo `app.modules.ts`
1. Localiza la sección de `@NgModule()`

> 1. Estando en la sección de `@NgModule()` remplaza el Array de `imports` con las siguientes lineas de código:

Reviewer: ```suggestion
1. Estando en la sección de `@NgModule()` remplaza el arreglo de `imports` con las siguientes líneas de código:
```

   <code-example header="src/app/app.module.ts" path="router-tutorial/src/app/app.module.ts" region="import-basic"></code-example>

> El fragmento de código anterior añade `RouterModules` al Array de `imports`. A continuación, el código usa el método `forRoot()` del `RouterModule` para definir tus dos rutas. Este método toma un Array de objetos JavaScript, con cada objeto definiendo las propiedades de una ruta. El método `forRoot()` asegura que su aplicación solo instancia un `RouterModule`. Para obtener más información, consulte [Singleton Services](/guide/singleton-services#forroot-and-the-router).

Reviewer: ```suggestion
El fragmento de código anterior añade `RouterModules` al arreglo de `imports`. A continuación, el código usa el método `forRoot()` del `RouterModule` para definir tus dos rutas. Este método toma un arreglo de objetos JavaScript, con cada objeto definiendo las propiedades de una ruta. El método `forRoot()` asegura que tu aplicación solo instancia un `RouterModule`. Para obtener más información consulta [Singleton Services](/guide/singleton-services#forroot-and-the-router).
```

## Actualiza tu componente con `router-outlet`

> Hasta este punto, has definido dos rutas para la aplicación. Sin embargo, la aplicación todavía tiene los componentes `crisis-center` y `heroes-list` codificados en su plantilla `app.component.html`. Para que sus rutas funcionen, necesita actualizar su plantilla para cargar dinámicamente un componente basado en la ruta URL.

Reviewer: ```suggestion
Hasta este punto, has definido dos rutas para la aplicación. Sin embargo, la aplicación todavía tiene los componentes `crisis-center` y `heroes-list` codificados en su plantilla `app.component.html`. Para que tus rutas funcionen, necesita actualizar tu plantilla para cargar dinámicamente un componente basado en la ruta URL.
```

> Para implementar esta funcionalidad, agregue la directiva `router-outlet` a su archivo de plantilla de acuerdo a los siguientes pasos:

Reviewer: ```suggestion
Para implementar esta funcionalidad, agregua la directiva `router-outlet` a tu archivo de plantilla de acuerdo a los siguientes pasos:
```

> 1.  Desde su editor de código abra el archivo `app.component.html`.

Reviewer: ```suggestion
1.  Desde tu editor de código abre el archivo `app.component.html`.
```
